Make EmptySet the instance rather than the class

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

Closes #16575
Fixes #16337

#### Brief description of what is fixed or changed

Makes EmptySet (as imported from sympy) be the singleton instance of the emptyset rather than the class. This brings EmptySet into line with other classes like Reals.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* sets
    * EmptySet (as imported from sympy) is now the singleton instance of the empty set rather than the class.
<!-- END RELEASE NOTES -->

dnabanita7 committed:
Changed importing of EmptySet

sympified EmptySet

dnabanita7 committed:
Updated object type to class type

changed EmptySet() to EmptySet.__class__() to get the class type rather than object
type after changed in __init__.py of sets.sets

dnabanita7 committed:
updated some changes

changes

dnabanita7 committed:
changes to switch types

changed object type of EmptySet to class type

dnabanita7 committed:
changed instance of EmptySet

to get the class of EmptySet we use type(EmptySet)

oscarbenjamin committed:
Resolve merge conflict hangover

oscarbenjamin committed:
Remove parens as EmptySet is the instance

oscarbenjamin committed:
Remove parens as EmptySet is the instance

oscarbenjamin commented:
I've hit a snag here which is that this change causes the symengine tests to fail. I think that's because symengine is importing and using EmptySet as a class rather than an instance. This can be fixed in both sympy and smyengine but changing either independently will fail.

How can we achieve the change simultaneously in both?

asmeurer commented:
CC @isuruf

asmeurer commented on sympy/utilities/tests/test_wester.py:
```diff
@@ -966,7 +966,7 @@ def test_M19():
 
 def test_M20():
-    assert solveset(sqrt(x**2 + 1) - x + 2, x) == EmptySet()
+    assert solveset(sqrt(x**2 + 1) - x + 2, x) == EmptySet.__class__()
```
Shouldn't this just be `EmptySet`?

oscarbenjamin replied:
Yes...

asmeurer commented:
Our end goal should be to do

```
diff --git a/sympy/core/tests/test_singleton.py b/sympy/core/tests/test_singleton.py
index 4eb24dae83..4bc47ceca7 100644
--- a/sympy/core/tests/test_singleton.py
+++ b/sympy/core/tests/test_singleton.py
@@ -84,4 +84,4 @@ def test_names_in_namespace():
             continue
 
         # Use is here because of complications with ==
-        assert any(getattr(S, name) is i or type(getattr(S, name)) is i for i in d.values()), name
+        assert any(getattr(S, name) is i for i in d.values()), name

```

Maybe we should fix this at the metaclass level somehow.

isuruf commented:
I can make SymEngine work with both SymPy release and the new version depending on SymPy's version string.

isuruf commented:
Let me know if this PR will be merged in SymPy 1.5 and if so I'll add the version checking for `>1.4` or not.

oscarbenjamin commented:
@isuruf the other problem is that I can't merge this in SymPy without disabling the symengine tests.

oscarbenjamin committed:
Use EmptySet not EmptySet.__class__()

isuruf commented on sympy/sets/sets.py:
```diff
@@ -466,15 +466,16 @@ def powerset(self):
+        >>> from sympy import FiniteSet, EmptySet
+        >>> A = EmptySet
         >>> A.powerset()
         {EmptySet()}
```
This printing is needs to be fixed, right?

oscarbenjamin replied:
That's fixed now. Some other doctests needed changing.

oscarbenjamin committed:
Make EmptySequence the instance not the class

isuruf committed:
Use symengine 0.5.*

oscarbenjamin committed:
Mkae EmptySet print as EmptySet

oscarbenjamin commented on sympy/printing/pretty/pretty_symbology.py:
```diff
@@ -509,6 +509,7 @@ def xsym(sym):
     'Intersection':            U('INTERSECTION'),
     'Ring':                    U('RING OPERATOR'),
     'Modifier Letter Low Ring':U('Modifier Letter Low Ring'),
+    'EmptySequence':           'EmptySequence',
```
Maybe there's a better unicode symbol for this

oscarbenjamin committed:
Revert inadvertant change from rebase

oscarbenjamin committed:
Change EmptySet() -> EmptySet

oscarbenjamin commented:
> Our end goal should be to do

That's done now. The only singletons that failed that test were EmptySet and EmptySequence and they are both changed here.

Having looked through this code I think it's very confusing having the same name for the class and the instance. It also creates import problems. Right now with this PR these give different objects:
```python
from sympy.sets import EmptySet      # instance
from sympy.sets.sets import EmptySet # class
```
I think it would be better to change the class to `EmptySetClass` and have `EmptySet` only refer to the instance. Then both can be importable from `sympy.sets.sets` and `sympy.sets.__init__.py` can just import EmptySet rather than this awkwardness:
https://github.com/sympy/sympy/blob/4b93506a88cb924785a0e1dc98bb15205d2d45b8/sympy/sets/__init__.py#L8-L15

oscarbenjamin committed:
Change EmptySet() to EmptySet

oscarbenjamin commented:
@isuruf there are new symengine failures that look like:
```
 sympy/physics/mechanics/tests/test_rigidbody.py:test_pendulum_angular_momentum 
Traceback (most recent call last):
  File "/home/travis/build/sympy/sympy/sympy/utilities/runtests.py", line 1315, in _timeout
    function()
  File "/home/travis/build/sympy/sympy/sympy/physics/mechanics/tests/test_rigidbody.py", line 100, in test_pendulum_angular_momentum
    I = inertia(R1, 0, m * a**2 / 3, m * a**2 / 3)
  File "/home/travis/build/sympy/sympy/sympy/physics/mechanics/functions.py", line 86, in inertia
    ol = sympify(ixx) * (frame.x | frame.x)
  File "symengine_wrapper.pyx", line 826, in symengine.lib.symengine_wrapper.Basic.__mul__
  File "symengine_wrapper.pyx", line 554, in symengine.lib.symengine_wrapper._sympify
  File "symengine_wrapper.pyx", line 436, in symengine.lib.symengine_wrapper.sympy2symengine
  File "/home/travis/build/sympy/sympy/sympy/physics/vector/dyadic.py", line 116, in __eq__
    other = _check_dyadic(other)
  File "/home/travis/build/sympy/sympy/sympy/physics/vector/dyadic.py", line 557, in _check_dyadic
    raise TypeError('A Dyadic must be supplied')
TypeError: A Dyadic must be supplied
```

oscarbenjamin commented:
This PR is now finished but blocked by the symengine tests

isuruf commented:
This is fixed in symengine master. I'll make a release soon.

isuruf commented on sympy/solvers/solveset.py:
```diff
@@ -843,7 +843,7 @@ def solve_decomposition(f, symbol, domain):
                 iter_iset = y_s.args
 
-            elif isinstance(y_s, EmptySet):
+            elif isinstance(y_s, type(EmptySet)):
```
Shouldn't this be `y_s is EmptySet`?

oscarbenjamin replied:
Yes

isuruf commented on sympy/solvers/solveset.py:
```diff
@@ -3395,7 +3395,7 @@ def nonlinsolve(system, *symbols):
         # positive dimensional system
         res = _handle_positive_dimensional(polys, symbols, denominators)
-        if isinstance(res, EmptySet) and any(not p.domain.is_Exact for p in polys):
+        if isinstance(res, type(EmptySet)) and any(not p.domain.is_Exact for p in polys):
```
Same here

oscarbenjamin replied:
Yes. I've checked that there are no other cases using `type(EmptySet)`.

oscarbenjamin committed:
Use is EmptySet rather than isinstance

isuruf commented:
I'll have symengine conda packages ready tomorrow.

isuruf committed:
SymEngine fixes

I'm not sure why these changes are unnecessary. One possible reason
could be that symengine now supports more classes and therefore
there are more symengine objects and comparing dictionary keys
fails because sometimes symengine objects are in dictionary keys
where it was sympy objects before.

isuruf committed:
Use symengine 0.5.1

oscarbenjamin commented:
Thanks very much @isuruf. This is now ready for review/merge.

jksuom commented:
This looks good to me.
